File: clients/android/app/src/main/java/us/i3u/hermesstudio/ui/theme/CoreHubTokens.kt
package us.i3u.hermesstudio.ui.theme

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * The single source of design values for Core Hub Mobile.
 *
 * Every number here is copied from docs/mobile/DESIGN-SPEC.md, which in turn is
 * extracted from the web client (`packages/client/src/styles/variables.scss`).
 * Screens must read these tokens (directly or through Material roles mapped in
 * [CoreHubTheme]) instead of carrying their own colours, radii or sizes.
 */
object CoreHubTokens {

    /** "Pure Ink": a greyscale base palette with no hue, in a light and a dark edition. */
    @Immutable
    data class Palette(
        val bgPrimary: Color,
        val bgSecondary: Color,
        val bgSidebar: Color,
        val bgCard: Color,
        val bgCardHover: Color,
        val bgInput: Color,
        /** The composer card; the same as [bgInput] in light mode, one step lighter in dark mode. */
        val bgComposer: Color,
        val border: Color,
        val borderLight: Color,
        val accent: Color,
        val accentHover: Color,
        val accentMuted: Color,
        val textPrimary: Color,
        val textSecondary: Color,
        val textMuted: Color,
        val success: Color,
        val error: Color,
        val warning: Color,
        val info: Color,
        val msgUser: Color,
        val msgAssistant: Color,
        val codeBg: Color,
        val textOnAccent: Color,
        /** Launch/splash ground and system bars. */
        val splash: Color,
        val isDark: Boolean,
    ) {
        /** hover = accent @ 6 %. */
        val hover: Color get() = accent.copy(alpha = Alpha.HOVER)

        /** active/selected = accent @ 12 %. */
        val selected: Color get() = accent.copy(alpha = Alpha.SELECTED)

        /** Idle input border = accent @ 18 %. */
        val inputBorder: Color get() = accent.copy(alpha = Alpha.INPUT_BORDER)

        /** Hovered input border = accent @ 32 %. */
        val inputBorderHover: Color get() = accent.copy(alpha = Alpha.INPUT_BORDER_HOVER)

        /** Text selection = accent @ 30 %. */
        val textSelection: Color get() = accent.copy(alpha = Alpha.TEXT_SELECTION)

        /** Error rows sit on error @ 6 %. */
        val errorSurface: Color get() = error.copy(alpha = Alpha.HOVER)

        /** Category tag background: #7f7f7f @ 12 %. */
        val tagBackground: Color get() = Color(0xFF7F7F7F).copy(alpha = Alpha.SELECTED)

        /** Segmented-control track: accent @ 5 %. */
        val segmentTrack: Color get() = accent.copy(alpha = Alpha.SEGMENT_TRACK)
    }

    val Light = Palette(
        bgPrimary = Color(0xFFFAFAFA),
        bgSecondary = Color(0xFFF0F0F0),
        bgSidebar = Color(0xFFF5F5F5),
        bgCard = Color(0xFFFFFFFF),
        bgCardHover = Color(0xFFFAFAFA),
        bgInput = Color(0xFFFFFFFF),
        bgComposer = Color(0xFFFFFFFF),
        border = Color(0xFFE0E0E0),
        borderLight = Color(0xFFEBEBEB),
        accent = Color(0xFF333333),
        accentHover = Color(0xFF1A1A1A),
        accentMuted = Color(0xFF888888),
        textPrimary = Color(0xFF1A1A1A),
        textSecondary = Color(0xFF666666),
        textMuted = Color(0xFF999999),
        success = Color(0xFF2E7D32),
        error = Color(0xFFC62828),
        warning = Color(0xFFF57F17),
        info = Color(0xFF4A90D9),
        msgUser = Color(0xFFF5F5F5),
        msgAssistant = Color(0xFFF5F5F5),
        codeBg = Color(0xFFF4F4F4),
        textOnAccent = Color(0xFFFFFFFF),
        splash = Color(0xFFF7F7F4),
        isDark = false,
    )

    val Dark = Palette(
        bgPrimary = Color(0xFF1A1A1A),
        bgSecondary = Color(0xFF252525),
        bgSidebar = Color(0xFF202020),
        bgCard = Color(0xFF2A2A2A),
        bgCardHover = Color(0xFF333333),
        bgInput = Color(0xFF2A2A2A),
        bgComposer = Color(0xFF333333),
        border = Color(0xFF3A3A3A),
        borderLight = Color(0xFF333333),
        accent = Color(0xFFE0E0E0),
        accentHover = Color(0xFFF5F5F5),
        accentMuted = Color(0xFF888888),
        textPrimary = Color(0xFFE0E0E0),
        textSecondary = Color(0xFFA0A0A0),
        textMuted = Color(0xFF888888),
        success = Color(0xFF66BB6A),
        error = Color(0xFFEF5350),
        warning = Color(0xFFFFB74D),
        info = Color(0xFF6BA3D6),
        msgUser = Color(0xFF292929),
        msgAssistant = Color(0xFF292929),
        codeBg = Color(0xFF1E1E1E),
        textOnAccent = Color(0xFF1A1A1A),
        splash = Color(0xFF1A1A1A),
        isDark = true,
    )

    /** State formulas, as alpha applied to the accent (or error) colour. */
    object Alpha {
        const val SEGMENT_TRACK = 0.05f
        const val HOVER = 0.06f
        const val SELECTED = 0.12f
        const val INPUT_BORDER = 0.18f
        const val TEXT_SELECTION = 0.30f
        const val INPUT_BORDER_HOVER = 0.32f

        /** The session-row delete ✕ is visible at half strength on touch. */
        const val DELETE_AFFORDANCE = 0.5f

        /** Unread halo around the 6 px dot. */
        const val UNREAD_HALO = SELECTED
    }

    /** Type ramp. Inter/system sans for text, JetBrains Mono/monospace for code. */
    object Type {
        val sans: FontFamily = FontFamily.SansSerif
        val mono: FontFamily = FontFamily.Monospace

        /** Base size; the web lets the user adjust 12–20. */
        val base: TextUnit = 14.sp
        val bodyLineHeight: TextUnit = 22.sp // 14 × 1.6
        val message: TextUnit = 14.sp
        val messageLineHeight: TextUnit = 23.sp // 14 × 1.65

        /** Page and chat titles: 16 / 600. */
        val title: TextUnit = 16.sp
        val titleWeight: FontWeight = FontWeight.SemiBold

        /** Sidebar nav item. */
        val navItem: TextUnit = 14.sp

        /** Page-sidebar tab and session title. */
        val navTab: TextUnit = 13.sp
        val sessionTitle: TextUnit = 13.sp

        /** Author name above a message. */
        val author: TextUnit = 12.sp

        /** Time, meta, tool line. */
        val meta: TextUnit = 11.sp

        /** Group header: 10 / 600 uppercase, +0.5 letter-spacing. */
        val groupHeader: TextUnit = 10.sp
        val groupHeaderWeight: FontWeight = FontWeight.SemiBold
        val groupHeaderLetterSpacing: TextUnit = 0.5.sp

        val categoryTag: TextUnit = 10.sp

        /** Code: 13 / 1.5 mono. */
        val code: TextUnit = 13.sp
        val codeLineHeight: TextUnit = 20.sp

        /** Thinking block: 13 italic at 85 % opacity. */
        val thinking: TextUnit = 13.sp
        const val thinkingAlpha: Float = 0.85f

        /** Inputs never render below 16 sp on phones (no zoom). */
        val inputMin: TextUnit = 16.sp

        val selectedWeight: FontWeight = FontWeight.Medium
    }

    /** Corner radii. */
    object Radius {
        /** Buttons, nav items, tool rows, code block, session row. */
        val small: Dp = 6.dp

        /** Naive default; tool summary header. */
        val medium: Dp = 8.dp

        /** Message bubble. */
        val bubble: Dp = 10.dp

        /** Cards: sidebar, session list, chat main. */
        val card: Dp = 14.dp

        /** Composer card. */
        val composer: Dp = 18.dp

        /** Pills. */
        val pill: Dp = 999.dp

        /** Category tag, workspace chip. */
        val tag: Dp = 4.dp

        /** Segmented tab. */
        val segment: Dp = 5.dp
    }

    /**
     * Shadows. Compose has no blur/offset shadow primitive on Material
     * surfaces; the elevation approximates "0 8 24 rgba(0,0,0,.10)" etc.
     */
    object Shadow {
        val card: Dp = 8.dp
        val composer: Dp = 8.dp
        val composerFocused: Dp = 10.dp
        val cardAlphaLight = 0.10f
        val composerAlphaLight = 0.08f
        val composerAlphaDark = 0.32f
        val focusedAlpha = 0.11f
    }

    /** Durations and layout metrics. */
    object Metrics {
        const val transitionFastMs = 150
        const val transitionMs = 250

        /** Drawer slide. */
        const val drawerSlideMs = 250
        const val scrimAlpha = 0.40f

        val sidebarWidth: Dp = 240.dp
        val sidebarCollapsed: Dp = 64.dp

        /** Off-canvas drawer on phones (sidebar width plus its padding). */
        val drawerWidth: Dp = 300.dp
        val headerHeight: Dp = 60.dp
        val breakpoint: Dp = 768.dp

        val segmentHeight: Dp = 30.dp
        val sessionRowPaddingV: Dp = 8.dp
        val sessionRowPaddingH: Dp = 10.dp
        val pinSize: Dp = 11.dp
        val unreadDot: Dp = 6.dp
        val agentAvatar: Dp = 18.dp
        val profileChipAvatar: Dp = 16.dp
        val groupChevron: Dp = 10.dp
        val messageAvatar: Dp = 22.dp
        val actionButton: Dp = 24.dp
        val composerButton: Dp = 30.dp
        val composerMinHeight: Dp = 150.dp
        val contextBarWidth: Dp = 42.dp
        val contextBarHeight: Dp = 4.dp
        val workspaceIcon: Dp = 12.dp

        /**
         * Bottom sheets (the composer's attachment sheet first): a 36 × 4
         * handle, a 20 dp gutter, rows at least 56 dp tall so a thumb never
         * misses, a 40 dp icon tile holding a 20 dp icon, and an 18 dp
         * trailing chevron.
         */
        val sheetHandleWidth: Dp = 36.dp
        val sheetHandleHeight: Dp = 4.dp
        val sheetPadding: Dp = 20.dp
        val sheetRowMinHeight: Dp = 56.dp
        val sheetRowGap: Dp = 14.dp
        val sheetIconTile: Dp = 40.dp
        val sheetIcon: Dp = 20.dp
        val sheetTrailingIcon: Dp = 18.dp
        val sheetButtonHeight: Dp = 44.dp

        /** Long-press before the session context menu opens. */
        const val longPressMs = 500L

        /** Session list "Recent" group defaults and bounds. */
        const val recentDefault = 10
        const val recentMin = 1
        const val recentMax = 100

        /** Context indicator turns amber above this share. */
        const val contextWarnRatio = 0.80f
        val contextWarn: Color = Color(0xFFE8A735)
    }
}

val LocalCoreHubPalette = staticCompositionLocalOf { CoreHubTokens.Light }

/** Shorthand for the palette in effect. */
object CoreHub {
    val palette: CoreHubTokens.Palette
        @Composable
        @ReadOnlyComposable
        get() = LocalCoreHubPalette.current
}

File: clients/ios/HermesStudio/Theme/CoreHubTokens.swift
import SwiftUI
import UIKit

/// Single source of the Core Hub design system on iOS.
/// Values come from `docs/mobile/DESIGN-SPEC.md` (extracted from the web
/// client's `variables.scss` / `theme.ts`). Never scatter raw colours, sizes
/// or radii in views; add a token here instead.
enum CoreHubTokens {
    // MARK: Palette ("Pure Ink", greyscale)

    /// Raw hex values so tests can assert the exact spec without rendering.
    struct PaletteHex: Equatable {
        let bgPrimary: UInt32
        let bgSecondary: UInt32
        let bgSidebar: UInt32
        let bgCard: UInt32
        let bgCardHover: UInt32
        let bgInput: UInt32
        /// The composer card background (web: input in light, #333333 in dark).
        let bgComposer: UInt32
        let border: UInt32
        let borderLight: UInt32
        let accent: UInt32
        let accentHover: UInt32
        let accentMuted: UInt32
        let textPrimary: UInt32
        let textSecondary: UInt32
        let textMuted: UInt32
        let textOnAccent: UInt32
        let success: UInt32
        let error: UInt32
        let warning: UInt32
        let info: UInt32
        let msgUser: UInt32
        let msgAssistant: UInt32
        let codeBackground: UInt32
        /// Splash / theme colour (also the app icon background).
        let splash: UInt32
        /// Context indicator above 80 %.
        let contextWarning: UInt32

        static let light = PaletteHex(
            bgPrimary: 0xFAFAFA, bgSecondary: 0xF0F0F0, bgSidebar: 0xF5F5F5, bgCard: 0xFFFFFF, bgCardHover: 0xFAFAFA,
            bgInput: 0xFFFFFF, bgComposer: 0xFFFFFF, border: 0xE0E0E0, borderLight: 0xEBEBEB,
            accent: 0x333333, accentHover: 0x1A1A1A, accentMuted: 0x888888,
            textPrimary: 0x1A1A1A, textSecondary: 0x666666, textMuted: 0x999999, textOnAccent: 0xFFFFFF,
            success: 0x2E7D32, error: 0xC62828, warning: 0xF57F17, info: 0x4A90D9,
            msgUser: 0xF5F5F5, msgAssistant: 0xF5F5F5, codeBackground: 0xF4F4F4, splash: 0xF7F7F4, contextWarning: 0xE8A735
        )

        static let dark = PaletteHex(
            bgPrimary: 0x1A1A1A, bgSecondary: 0x252525, bgSidebar: 0x202020, bgCard: 0x2A2A2A, bgCardHover: 0x333333,
            bgInput: 0x2A2A2A, bgComposer: 0x333333, border: 0x3A3A3A, borderLight: 0x333333,
            accent: 0xE0E0E0, accentHover: 0xF5F5F5, accentMuted: 0x888888,
            textPrimary: 0xE0E0E0, textSecondary: 0xA0A0A0, textMuted: 0x888888, textOnAccent: 0x1A1A1A,
            success: 0x66BB6A, error: 0xEF5350, warning: 0xFFB74D, info: 0x6BA3D6,
            msgUser: 0x292929, msgAssistant: 0x292929, codeBackground: 0x1E1E1E, splash: 0x1A1A1A, contextWarning: 0xE8A735
        )
    }

    /// Dynamic colours that follow the effective colour scheme (system, or
    /// the user's light/dark preference applied through `preferredColorScheme`).
    enum Palette {
        static let bgPrimary = dynamic(\.bgPrimary)
        static let bgSecondary = dynamic(\.bgSecondary)
        static let bgSidebar = dynamic(\.bgSidebar)
        static let bgCard = dynamic(\.bgCard)
        static let bgCardHover = dynamic(\.bgCardHover)
        static let bgInput = dynamic(\.bgInput)
        static let bgComposer = dynamic(\.bgComposer)
        static let border = dynamic(\.border)
        static let borderLight = dynamic(\.borderLight)
        static let accent = dynamic(\.accent)
        static let accentHover = dynamic(\.accentHover)
        static let accentMuted = dynamic(\.accentMuted)
        static let textPrimary = dynamic(\.textPrimary)
        static let textSecondary = dynamic(\.textSecondary)
        static let textMuted = dynamic(\.textMuted)
        static let textOnAccent = dynamic(\.textOnAccent)
        static let success = dynamic(\.success)
        static let error = dynamic(\.error)
        static let warning = dynamic(\.warning)
        static let info = dynamic(\.info)
        static let msgUser = dynamic(\.msgUser)
        static let msgAssistant = dynamic(\.msgAssistant)
        static let codeBackground = dynamic(\.codeBackground)
        static let splash = dynamic(\.splash)
        static let contextWarning = dynamic(\.contextWarning)

        /// State formulas from the spec, derived from `accent`.
        static var hover: Color { accent.opacity(Alpha.hover) }
        static var selected: Color { accent.opacity(Alpha.selected) }
        static var inputBorderIdle: Color { accent.opacity(Alpha.inputBorderIdle) }
        static var segmentTrack: Color { accent.opacity(Alpha.segmentTrack) }
        /// Category tag background: #7f7f7f @ 12 % in both schemes.
        static let categoryTag = Color(red: 0x7F / 255.0, green: 0x7F / 255.0, blue: 0x7F / 255.0).opacity(Alpha.categoryTag)

        private static func dynamic(_ key: KeyPath<PaletteHex, UInt32>) -> Color {
            Color(uiColor: UIColor { traits in
                let hex = traits.userInterfaceStyle == .dark ? PaletteHex.dark[keyPath: key] : PaletteHex.light[keyPath: key]
                return UIColor(coreHubHex: hex)
            })
        }
    }

    enum Alpha {
        static let hover = 0.06
        static let selected = 0.12
        static let inputBorderIdle = 0.18
        static let inputBorderHover = 0.32
        static let textSelection = 0.30
        static let segmentTrack = 0.05
        static let categoryTag = 0.12
        static let drawerScrim = 0.40
        static let unreadHalo = 0.12
        static let deleteAffordance = 0.50
        static let thinkingText = 0.85
    }

    // MARK: Typography (system sans / system mono)

    enum Typography {
        /// Device-local text scale (Settings › Display › Text size). Every
        /// token font is multiplied by it; `AppStore` keeps it in step with
        /// the stored preference and rebuilds the shell when it changes.
        static var scale: CGFloat = 1

        static let base: CGFloat = 14
        static let baseMinimum: CGFloat = 12
        static let baseMaximum: CGFloat = 20
        static let title: CGFloat = 16
        static let titleWeight: Font.Weight = .semibold
        static let navItem: CGFloat = 14
        static let sidebarTab: CGFloat = 13
        static let sessionTitle: CGFloat = 13
        static let author: CGFloat = 12
        static let meta: CGFloat = 11
        static let groupHeader: CGFloat = 10
        static let groupHeaderWeight: Font.Weight = .semibold
        static let groupHeaderTracking: CGFloat = 0.5
        static let categoryTag: CGFloat = 10
        static let code: CGFloat = 13
        static let thinking: CGFloat = 13
        static let messageBody: CGFloat = 14
        static let bodyLineHeight: CGFloat = 1.6
        static let messageLineHeight: CGFloat = 1.65
        static let codeLineHeight: CGFloat = 1.5
        /// Inputs never render below 16 pt on phones (no zoom on focus).
        static let inputMinimum: CGFloat = 16
        static let workspaceChip: CGFloat = 11

        /// Clamped so a large scale never breaks the 16 pt input minimum.
        static func scaled(_ size: CGFloat) -> CGFloat { (size * max(0.85, min(1.45, scale))).rounded() }

        static func font(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
            .system(size: scaled(size), weight: weight)
        }
        static func mono(_ size: CGFloat = code, weight: Font.Weight = .regular) -> Font {
            .system(size: scaled(size), weight: weight, design: .monospaced)
        }

        static var titleFont: Font { font(title, weight: titleWeight) }
        static var navItemFont: Font { font(navItem) }
        static var sidebarTabFont: Font { font(sidebarTab) }
        static var sessionTitleFont: Font { font(sessionTitle) }
        static var authorFont: Font { font(author) }
        static var metaFont: Font { font(meta) }
        static var groupHeaderFont: Font { font(groupHeader, weight: groupHeaderWeight) }
        static var categoryTagFont: Font { font(categoryTag) }
        static var bodyFont: Font { font(base) }
        static var messageFont: Font { font(messageBody) }
        static var thinkingFont: Font { font(thinking).italic() }
        static var codeFont: Font { mono() }
    }

    // MARK: Radii

    enum Radius {
        /// Buttons, nav items, tool rows, code blocks, session rows.
        static let button: CGFloat = 6
        /// Default control radius (tool summary header).
        static let control: CGFloat = 8
        static let bubble: CGFloat = 10
        /// Cards: sidebar, session list, chat main.
        static let card: CGFloat = 14
        static let composer: CGFloat = 18
        static let pill: CGFloat = 999
        /// Category tag, workspace chip.
        static let tag: CGFloat = 4
        static let segment: CGFloat = 5
    }

    // MARK: Shadows (CSS `0 y blur alpha`; SwiftUI radius ≈ blur / 2)

    struct ShadowSpec: Equatable {
        let opacity: Double
        let blur: CGFloat
        let y: CGFloat
        var radius: CGFloat { blur / 2 }
        var color: Color { Color.black.opacity(opacity) }
    }

    enum Shadow {
        static let card = ShadowSpec(opacity: 0.10, blur: 24, y: 8)
        static let composer = ShadowSpec(opacity: 0.08, blur: 28, y: 8)
        static let composerDark = ShadowSpec(opacity: 0.32, blur: 28, y: 8)
        static let focused = ShadowSpec(opacity: 0.11, blur: 32, y: 10)

        static func composer(for scheme: ColorScheme) -> ShadowSpec { scheme == .dark ? composerDark : composer }
    }

    // MARK: Motion

    enum Motion {
        static let fast: Double = 0.15
        static let normal: Double = 0.25
        static var drawer: Animation { .easeInOut(duration: normal) }
        static var quick: Animation { .easeInOut(duration: fast) }
    }

    // MARK: Layout

    enum Layout {
        static let sidebarWidth: CGFloat = 240
        static let sidebarCollapsedWidth: CGFloat = 64
        static let headerHeight: CGFloat = 60
        static let breakpoint: CGFloat = 768
        /// Off-canvas drawer: at most this wide, at most this fraction of the screen.
        static let drawerMaxWidth: CGFloat = 300
        static let drawerWidthFraction: CGFloat = 0.84
        static let edgeSwipeWidth: CGFloat = 18
        static let iconStroke: CGFloat = 1.8
        static let iconViewBox: CGFloat = 24
        static let railIcon: CGFloat = 20
        static let segmentHeight: CGFloat = 30
        static let sessionRowVertical: CGFloat = 8
        static let sessionRowHorizontal: CGFloat = 10
        static let sessionAvatar: CGFloat = 18
        static let profileChipAvatar: CGFloat = 16
        static let pinIcon: CGFloat = 11
        static let unreadDot: CGFloat = 6
        static let groupChevron: CGFloat = 10
        static let categoryTagMaxWidthFraction: CGFloat = 0.45
        static let assistantAvatar: CGFloat = 22
        static let userBubbleMaxFraction: CGFloat = 0.75
        static let assistantBubbleMaxFraction: CGFloat = 0.80
        static let bubblePaddingVertical: CGFloat = 10
        static let bubblePaddingHorizontal: CGFloat = 14
        static let composerMinHeight: CGFloat = 150
        static let composerButton: CGFloat = 30
        static let actionButton: CGFloat = 24
        static let contextBarWidth: CGFloat = 60
        static let contextBarWidthPhone: CGFloat = 42
        static let contextBarHeight: CGFloat = 4
        static let modelPillMaxWidth: CGFloat = 190
        static let toolSummaryMaxWidth: CGFloat = 520
        static let toolSummaryHeader: CGFloat = 30
        static let longPress: Double = 0.5
        static let recentDefault = 10
        static let recentMinimum = 1
        static let recentMaximum = 100
    }
}

extension UIColor {
    convenience init(coreHubHex hex: UInt32, alpha: CGFloat = 1) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: alpha
        )
    }
}

extension View {
    /// Applies one of the spec shadows.
    func coreHubShadow(_ spec: CoreHubTokens.ShadowSpec) -> some View {
        shadow(color: spec.color, radius: spec.radius, x: 0, y: spec.y)
    }

    /// Page background (bg.primary) behind every screen.
    func hermesBackground() -> some View { background(CoreHubTokens.Palette.bgPrimary.ignoresSafeArea()) }
}
